### PR TITLE
Various minor improvements on the console outputs for multiapps [Adding Magenta]

### PR DESCRIPTION
**Pull request by YaqiWang**
This makes the screen cleaner...

I do not know why I need a flush in `EigenExecutionerBase::printEigenvalue()`, otherwise the output is delayed randomly.

Also I make some of print-outs in magenta color to highlight the execution of multiapps.

Referring #5956.

### CODE REVIEW

File: framework/src/executioners/EigenExecutionerBase.C [hunk 1]
```diff
@@ -319,7 +319,7 @@ EigenExecutionerBase::inversePowerIteration(unsigned int min_iter,
 
     if (echo)
       _console << " ________________________________________________________________________________ "
-               << std::endl;
+               << std::endl << std::flush;
```
Comment (permcody) [inline framework/src/executioners/EigenExecutionerBase.C]: I don't think this is necessary. I'll have to take a look at what you are seeing. By the very definition of these two methods, you are flushing twice!

Comment (YaqiWang) [inline framework/src/executioners/EigenExecutionerBase.C]: Yes, possibly. I can give you a fancy input file which has five multiapps a little later. But the other one in `printEigenvalue` seems necessary.

Comment (YaqiWang) [inline framework/src/executioners/EigenExecutionerBase.C]: @permcody you can checkout my yak branch `coupled_reactor`. The input is under `yak/tests/actions/neutron_diffusion_cfem/master.i`. We can use it to fix all minor output issues.

**Comment by YaqiWang** [comment]
@aeslaughter I am good with the current improvement in `OutputWarehouse` but not sure if you are ok with it or not.

File: framework/src/outputs/OutputWarehouse.C [hunk 2]
```diff
@@ -178,6 +179,15 @@ OutputWarehouse::mooseConsole()
     _console_buffer.clear();
     _console_buffer.str("");
   }
+  else
```
Comment (aeslaughter) [inline framework/src/outputs/OutputWarehouse.C]: There should be an option to disable/enable this feature, in case things are being redirected to a file.

Comment (YaqiWang) [inline framework/src/outputs/OutputWarehouse.C]: ok.

**Comment by YaqiWang** [comment]
I'd like one of you take a look on that input to see if anything still can be improved.

**Comment by permcody** [comment]
Will do, might be slow to respond over the next week or slow. Half the team is gone and I'm busy doing audit stuff.

**Comment by YaqiWang** [comment]
Note that although I am referring #5956, all these minor fixes are not fixing minor issues listed in #5956.

**Comment by YaqiWang** [comment]
Not in hurry. Maybe Andrew will have a better fix along this line.

**Comment by YaqiWang** [comment]
This can be merged if @aeslaughter agreed.

File: framework/src/base/FEProblem.C [hunk 3]
```diff
@@ -515,12 +515,24 @@ void FEProblem::initialSetup()
   // HUGE NOTE: MultiApp initialSetup() MUST... I repeat MUST be _after_ restartable data has been restored
 
   // Call initialSetup on the MultiApps
+  bool has_multiapp = false;
+  if (_multi_apps(EXEC_LINEAR)[0].all().size() ||
```
Comment (permcody) [inline framework/src/base/FEProblem.C]: Why not just set this Boolean when you we add a Multiapp to the system? I could have sworn that we already had a Boolean for this...

Comment (YaqiWang) [inline framework/src/base/FEProblem.C]: really? I did not check carefully.

File: framework/src/outputs/OutputWarehouse.C [hunk 4]
```diff
@@ -178,6 +180,19 @@ OutputWarehouse::mooseConsole()
...
+  {
+    if (!_buffer_action_console_outputs)
+    {
+      // this will cause messages to console before its construction immediately flushed and cleared.
+      std::string message = _console_buffer.str();
+      if (_app.multiAppLevel() > 0)
+        MooseUtils::indentMessage(_app.name(), message);
+      Moose::out << message;
+      _console_buffer.clear();
+      _console_buffer.str("");
```
Comment (permcody) [inline framework/src/outputs/OutputWarehouse.C]: Are both of these lines necessary?

File: framework/src/outputs/Console.C [hunk 5]
```diff
@@ -587,13 +587,22 @@ Console::write(std::string message, bool indent /*=true*/)
...
   if (indent && _app.multiAppLevel() > 0)
     MooseUtils::indentMessage(_app.name(), message);
 
   // Write message to the screen
   if (_write_screen)
+  {
     Moose::out << message;
+    if (has_end_newline)
+      Moose::out << "\n";
+  }
```
Comment (permcody) [inline framework/src/outputs/Console.C]: This whole change doesn't make sense. So first you are looking at the last character in the string to see if it's a newline. Then you strip it off, then you indent, then you add it back. In the indent routine you took of the logic to not look at the final newline. At a glance this all looks like a net-zero change. Does the old "much more compact" logic not work?

File: framework/src/utils/MooseUtils.C [hunk 6]
```diff
@@ -355,8 +355,8 @@ indentMessage(const std::string & prefix, std::string & message, const char* col
   // The colored prefix
   std::string indent = color + prefix + ": " + COLOR_DEFAULT;
 
-  // Indent all lines after the first
-  pcrecpp::RE re("\n(?!\\Z)");
+  // Indent all lines
+  pcrecpp::RE re("\n");
```
Comment (permcody) [inline framework/src/utils/MooseUtils.C]: The corresponding logic to the comment above.

File: framework/src/multiapps/MultiApp.C [hunk 7]
```diff
@@ -124,9 +124,6 @@ MultiApp::MultiApp(const InputParameters & parameters):
 
   mooseAssert(_input_files.size() == 1 || _positions.size() == _input_files.size(), "Number of positions and input files are not the same!");
 
-  // Fill in the _positions vector
-  fillPositions();
```
Comment (permcody) [inline framework/src/multiapps/MultiApp.C]: This looks suspicious.  Do we not need this at all? I don't see a corresponding re-add of this removed line.

**Comment by permcody** [comment]
Most of this looks OK and I'm pleased that you got rid of the the double flush. I made a few comments though that I'd like you to address.

File: framework/src/multiapps/MultiApp.C (hunk 7 above)
Comment (YaqiWang) [inline framework/src/multiapps/MultiApp.C]: I think this is duplicated. If you look the full file, you will see.

File: framework/src/outputs/Console.C (hunk 5 above)
Comment (YaqiWang) [inline framework/src/outputs/Console.C]: looks like a net-zero change but it is not. You can see the difference with my test. It is possibly related with `_write_screen` but I have no idea on that.

File: framework/src/outputs/OutputWarehouse.C (hunk 4 above)
Comment (YaqiWang) [inline framework/src/outputs/OutputWarehouse.C]: See the other branch. I am copying these two lines. Not sure if it is necessary.

File: framework/src/utils/MooseUtils.C (hunk 6 above)
Comment (YaqiWang) [inline framework/src/utils/MooseUtils.C]: I actually try to move the code in Console.C to here but it does not work. There is something going on i am not aware of. That is why I want Andrew the original developer to take a look at.

File: framework/src/base/FEProblem.C (hunk 3 above)
Comment (YaqiWang) [inline framework/src/base/FEProblem.C]: Can you point to me where this bool is set?

**Comment by aeslaughter** [comment]
@YaqiWang Would you send a before and after screen shot of the output, I would like to get a visual for the changes you are making?

**Comment by YaqiWang** [comment]
Sure, I will send you the outputs before and after that ci.

File: framework/src/base/MooseApp.C [hunk 8]
```diff
@@ -117,6 +118,7 @@ MooseApp::MooseApp(InputParameters parameters) :
     _start_time_set(false),
     _start_time(0.0),
     _global_time_offset(0.0),
+    _output_warehouse(*this, getParam<bool>("buffer_action_console_outputs")),
```
Comment (aeslaughter) [inline framework/src/base/MooseApp.C]: For some reason, I don't like this option being passed in the constructor of the warehouse. I would rather see MooseApp call a method in the constructor:

```
_output_warehouse.bufferActionOutputsBeforeConstruction(getParam<bool>("buffer..."));
```

**Comment by aeslaughter** [comment]
@YaqiWang The only difference in the two outputs seems to be the absence of empty lines, which is good. So, if you address Cody's comment, who is more up-to-speed on the buffer related stuff, I am good.

File: framework/src/base/MooseApp.C (hunk 8 above)
Comment (YaqiWang) [inline framework/src/base/MooseApp.C]: will change it.

**Comment by YaqiWang** [comment]
I will try to remove the changes made in MooseUtils.C see if it affects anything. Maybe Console is special and we need a special treatment on the newlines.

File: framework/src/base/FEProblem.C (hunk 3 above)
Comment (YaqiWang) [inline framework/src/base/FEProblem.C]: I have found it, `_has_multiapps`.

**Comment by YaqiWang** [comment]
I do not like the changes I made in `MooseUitls.C`, but it is necessary to fix the empty line issue...

**Comment by permcody** [comment]
Are you finished making changes? If so, I'd like to take a look to see if I can understand why your MooseUtils change "seems" necessary.

**Comment by YaqiWang** [comment]
Yes

File: framework/src/outputs/Console.C (hunk 5 above)
Comment (permcody) [inline framework/src/outputs/Console.C]: Which test are you running to look at the change in output?

Comment (YaqiWang) [inline framework/src/outputs/Console.C]: Grab my yak branch `couple_reactor`. The test is under yak/tests/actions/neutron_diffusion_cfem/master.i.

File: framework/src/multiapps/MultiApp.C (hunk 7 above)
Comment (permcody) [inline framework/src/multiapps/MultiApp.C]: Definitely not duplicated! This should cause something to break. I'm investigating a little more.

Comment (YaqiWang) [inline framework/src/multiapps/MultiApp.C]: There are two lines with the same call. I think it is a copy-paste bug.

Comment (permcody) [inline framework/src/multiapps/MultiApp.C]: Found it!

**Comment by permcody** [comment]
Closing in favor of #6021.

**Pull request closed without merging**